apps/src/blockly/addons/cdoBlockSerializer.js

```diff
@@ -1,7 +1,25 @@
 import GoogleBlockly from 'blockly/core';
+import {procedureDefinitionTypes} from '../constants';
 
 const unknownBlockState = {type: 'unknown', enabled: false};
 
+// Sorting function, used by load()
+function sortBlocksByType(blockStates, prioritizedBlockTypes) {
+  // Create a copy of the blockStates array in case it's read-only
+  const copiedBlockStates = [...blockStates];
+
+  // This function sorts blocks of the specified types to the front of the list.
+  return copiedBlockStates.sort((a, b) => {
+    if (prioritizedBlockTypes.includes(a.type)) {
```

Comment:
This handles multiple block types because they always need to be in order from highest priority to lowest priority and `.includes` always searches front to back, right? Maybe worth adding a comment to that effect?

Reply:
Not actually. This handles multiple block types because we'll also need to sort behavior definition blocks to front of the list, once those are ready. The exact order does matter; as long as procedure definitions (of any type) are loaded before their associated callers, we should be good.

Reply:
I think my question is not, "Why do we need to support multiple block types?" but, "If we had a scenario where prioritizedblockTypes was an array like [blockTypeA, blockTypeB] indicating that all of type A needed to go first, all of type B needed to second, and then everything else goes last, why is it true that this code will work in this scenario?" Do you think that question makes sense?

Reply:
> If we had a scenario where prioritizedblockTypes was an array like [blockTypeA, blockTypeB] indicating that all of type A needed to go first, all of type B needed to second, and then everything else goes last, why is it true that this code will work in this scenario?"

I don't think it would work in that scenario. If it compared a block of type A and a block of type B, where both types are part of the `prioritizedblockTypes` list, block A would be positioned before block B. The order of the blocks doesn't matter if both are of a type we want to prioritize, so I think this is okay.

Reply:
Could we change the code to only support one prioritized type for now? I think this might clarify the purpose, and we don't anticipate a use case for anything else, do we? (Or change `prioritizedBlockTypes` to be an object instead of an array to indicate that there's no significance to the order of the `prioritizedBlockTypes` array?) @molly-moen, thoughts? (I might be over-indexing on this.)

Reply:
Open to suggestions here! FWIW, my next PR will add a second prioritized type (`'behavior_definition'`), which is why I went with a list of one rather than a single string for now. Both block types will need to be sorted to the front of the list. The order just impacts loading and adding procedure models. As long as the definition types are loaded before the call types, we should be good. So, it's not really "sort" as in "put them all in a specific order" so much as "filter out blocks of the specified types and put them first." Maybe I could refactor to use `.filter()` and then merge the two lists or something? Happy to change labels or add comments to help clarify this intended use!

Reply:
That makes sense! You're right; this strikes me as a partitioning problem more than a "sorting" problem. What about something like:

```
function partitionBlocks(blockStates, typesToPartitionFirst) {
    let first = [];
    let rest = [];

    blocks.forEach(block => {
        typesToPartitionFirst.includes(block.type) ? first.push(block) : rest.push(block);
    });

    return [...first, ...rest];
}
```

Reply:
I can see how the current name could be confusing, so switching it up/clearly commenting it makes sense. Maybe something like `topBlockTypes`? Although I also keep coming back to `prioritized`, the problem is it could mean "prioritized over all other blocks not in this list" (the intended definition), or "this list is in priority order" (unintended definition)

Reply:
I refactored the sorting here and in CdoXml to use @ebeastlake's suggested partitioning approach:
https://github.com/code-dot-org/code-dot-org/pull/52932/commits/4313d916fab117134b856bb88558f4e2a4a7b254

@molly-moen As of this commit, I'm still using "prioritized". Hopefully because we're no longer using a true `.sort()`, it's likely to falsely convey that the blocks are in "priority order".  PTAL and let me know if you think there is a better way to describe this.

Reply:
Looks good to me!